src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,174 @@
+        } else {
+            lastBeforeAdd.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkPositionIndex(index);
+        if (index == size) {
+            add(value);
+        } else {
+            linkBefore(value, getNode(index));
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkElementIndex(index);
+        return getNode(index).value;
     }
 
     @Override
-    public T set(T value, int index) {
-        return null;
+    public T set(T element, int index) {
+        checkElementIndex(index);
+        Node<T> currentNode = getNode(index);
+        T oldValue = currentNode.value;
+        currentNode.value = element;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkElementIndex(index);
+        return unlink(getNode(index));
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (Node<T> x = head; x != null; x = x.next) {
+                if (x.value == null) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        } else {
+            for (Node<T> x = head; x != null; x = x.next) {
+                if (object.equals(x.value)) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private T unlink(Node<T> nodeToUnlink) {
+        final T element = nodeToUnlink.value;
+        final Node<T> nextNode = nodeToUnlink.next;
+        final Node<T> previousNode = nodeToUnlink.prev;
+
+        if (previousNode == null) {
+            head = nextNode;
+        } else {
+            previousNode.next = nextNode;
+            nodeToUnlink.prev = null;
+        }
+        if (nextNode == null) {
+            tail = previousNode;
+        } else {
+            nextNode.prev = previousNode;
+            nodeToUnlink.next = null;
+        }
+        nodeToUnlink.value = null;
+        size--;
+        return element;
+    }
+
+    private void linkBefore(T value, Node<T> displacedNode) {
+        final Node<T> previousOne = displacedNode.prev;
+        final Node<T> newNode = new Node<>(previousOne, value, displacedNode);
+        displacedNode.prev = newNode;
+        if (previousOne == null) {
+            head = newNode;
+        } else {
+            previousOne.next = newNode;
+        }
+        size++;
+    }
+
+    private Node<T> getNode(int index) {
+        if (index < (size >> 1)) {
```

Comment:
This oen better for readability
```suggestion
        if (index < (size / 2)) {
```